cpp/src/arrow/dataset/discovery.cc

```diff
@@ -134,8 +135,30 @@ Result<std::shared_ptr<DatasetFactory>> FileSystemDatasetFactory::Make(
 Result<std::shared_ptr<DatasetFactory>> FileSystemDatasetFactory::Make(
     std::shared_ptr<fs::FileSystem> filesystem, const std::vector<fs::FileInfo>& files,
     std::shared_ptr<FileFormat> format, FileSystemFactoryOptions options) {
+  // Discover files in directories and globs
+  std::vector<fs::FileInfo> discovered_files;
+  for (const auto& file : files) {
+    if (file.IsDirectory()) {
+      fs::FileSelector file_selector;
+      file_selector.base_dir = file.dir_name();
+      file_selector.recursive = true;
+      ARROW_ASSIGN_OR_RAISE(auto folder_files,
+                            filesystem->GetFileInfo(file_selector));
+      std::move(folder_files.begin(), folder_files.end(), std::back_inserter(discovered_files));
+    } else if (file.IsGlob()) {
+      glob_t glob_result;
+      memset(&glob_result, 0, sizeof(glob_result)); 
```

Comment:
```suggestion
      glob_t glob_result{};
```
Minor nit: As of C++11 you can use `{}` which forces value initialization (e.g. zero-filled) instead of default initialization.